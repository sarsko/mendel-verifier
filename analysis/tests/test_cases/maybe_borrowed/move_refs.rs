#[analyzer::run]
fn move_unit_shared_ref() {
    let unit = ();
    let unit_ref = &unit;
    let tuple = (unit_ref, unit_ref);
}

#[analyzer::run]
fn move_unit_mutable_ref() {
    let mut unit1 = ();
    let mut unit2 = ();
    let unit_ref1 = &mut unit1;
    let unit_ref2 = &mut unit2;
    let tuple = (unit_ref1, unit_ref2);
}

fn main() {}
