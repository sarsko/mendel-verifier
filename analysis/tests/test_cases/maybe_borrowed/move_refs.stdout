Analyzing file $DIR/move_refs.rs using MaybeBorrowedAnalysis...
Result for function move_unit_shared_ref():
{
  "bb0": [
    [
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: StorageLive(_1)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: _1 = ()"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: FakeRead(ForLet(None), _1)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: StorageLive(_2)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: _2 = &_1"
      ],
      [
        "state:",
        {
          "frozen": [
            "_1"
          ],
          "blocked": []
        },
        "statement: FakeRead(ForLet(None), _2)"
      ],
      [
        "state:",
        {
          "frozen": [
            "_1"
          ],
          "blocked": []
        },
        "statement: StorageLive(_3)"
      ],
      [
        "state:",
        {
          "frozen": [
            "_1"
          ],
          "blocked": []
        },
        "statement: StorageLive(_4)"
      ],
      [
        "state:",
        {
          "frozen": [
            "_1"
          ],
          "blocked": []
        },
        "statement: _4 = _2"
      ],
      [
        "state:",
        {
          "frozen": [
            "_1"
          ],
          "blocked": []
        },
        "statement: StorageLive(_5)"
      ],
      [
        "state:",
        {
          "frozen": [
            "_1"
          ],
          "blocked": []
        },
        "statement: _5 = _2"
      ],
      [
        "state:",
        {
          "frozen": [
            "_1"
          ],
          "blocked": []
        },
        "statement: _3 = (move _4, move _5)"
      ],
      [
        "state:",
        {
          "frozen": [
            "_1"
          ],
          "blocked": []
        },
        "statement: StorageDead(_5)"
      ],
      [
        "state:",
        {
          "frozen": [
            "_1"
          ],
          "blocked": []
        },
        "statement: StorageDead(_4)"
      ],
      [
        "state:",
        {
          "frozen": [
            "_1"
          ],
          "blocked": []
        },
        "statement: FakeRead(ForLet(None), _3)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: _0 = const ()"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: StorageDead(_3)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: StorageDead(_2)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: StorageDead(_1)"
      ]
    ],
    "state before terminator:",
    {
      "frozen": [],
      "blocked": []
    },
    "terminator: return",
    {}
  ]
}
Result for function move_unit_mutable_ref():
{
  "bb0": [
    [
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: StorageLive(_1)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: _1 = ()"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: FakeRead(ForLet(None), _1)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: StorageLive(_2)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: _2 = ()"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: FakeRead(ForLet(None), _2)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: StorageLive(_3)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: _3 = &mut _1"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": [
            "_1"
          ]
        },
        "statement: FakeRead(ForLet(None), _3)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": [
            "_1"
          ]
        },
        "statement: StorageLive(_4)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": [
            "_1"
          ]
        },
        "statement: _4 = &mut _2"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": [
            "_1",
            "_2"
          ]
        },
        "statement: FakeRead(ForLet(None), _4)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": [
            "_1",
            "_2"
          ]
        },
        "statement: StorageLive(_5)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": [
            "_1",
            "_2"
          ]
        },
        "statement: StorageLive(_6)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": [
            "_1",
            "_2"
          ]
        },
        "statement: _6 = move _3"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": [
            "_1",
            "_2"
          ]
        },
        "statement: StorageLive(_7)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": [
            "_1",
            "_2"
          ]
        },
        "statement: _7 = move _4"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": [
            "_1",
            "_2"
          ]
        },
        "statement: _5 = (move _6, move _7)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": [
            "_1",
            "_2"
          ]
        },
        "statement: StorageDead(_7)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": [
            "_1",
            "_2"
          ]
        },
        "statement: StorageDead(_6)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": [
            "_1",
            "_2"
          ]
        },
        "statement: FakeRead(ForLet(None), _5)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: _0 = const ()"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: StorageDead(_5)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: StorageDead(_4)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: StorageDead(_3)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: StorageDead(_2)"
      ],
      [
        "state:",
        {
          "frozen": [],
          "blocked": []
        },
        "statement: StorageDead(_1)"
      ]
    ],
    "state before terminator:",
    {
      "frozen": [],
      "blocked": []
    },
    "terminator: return",
    {}
  ]
}
